### ISSUE TITLE
"Soft tabs" don't work (tab and enter keys add tabs instead of spaces)

### ISSUE THREAD

**Issue by mik01aj** [atom/atom#5497]
![screenshot 2015-02-11 11 59 46](https://cloud.githubusercontent.com/assets/3027415/6145763/bc1c2bc4-b1e5-11e4-8102-953c9a83beb2.png)

In this situation, when I press enter, I would expect leading spaces to be added to the new line. But when I do, tabs are added instead:

![screenshot 2015-02-11 11 59 59](https://cloud.githubusercontent.com/assets/3027415/6145765/be94e300-b1e5-11e4-8a1e-6568571d4e50.png)

Additionally, when I press the tab key, it adds tabs, not spaces.

Here are my preferences - the relevant part from `config.cson`:

    editor:
        showInvisibles: true
        showIndentGuide: true
        preferredLineLength: 100
        tabLength: 4
        invisibles: {}
        fontSize: 12

And a part of preferences screen (just in case): https://www.dropbox.com/s/vj64h58mg43dsr8/Screenshot%202015-02-11%2012.06.26.png?dl=0

I'm using Atom `0.177.0`.

**Comment by izuzak** [comment]
Can you try running the "Editor: Toggle Soft Tabs" command from the command palette while that file is open, and see if that helps?

**Comment by mik01aj** [comment]
Oh yes, toggling "Editor: Toggle Soft Tabs" from the command palette works great. So what about the preferences: maybe it is just a default for newly-opened editor tabs? If so, I believe that this should be documented below the checkbox in preferences.

**Comment by izuzak** [comment]
> So what about the preferences: maybe it is just a default for newly-opened editor tabs? 

No, there's some auto-detection:

https://github.com/atom/atom/blob/master/src/text-editor.coffee#L2350-L2369

Any chance the file has a tab as the first character of the first non-comment line? Or any chance you already triggered the "Editor: Toggle Soft Tabs" command previously? Can you share a link to the file in question or share it in a gist?

See https://github.com/atom/atom/issues/5192 for a similar issue (this should be fixed when https://github.com/atom/atom/issues/3719 is fixed).

**Comment by mik01aj** [comment]
I think it might be related to js-formatter plugin which converted all my spaces to tabs, and only after configuring it I got the spaces there. So at the time when I was opening the file it had tabs.

I would love to have some status bar icon showing the current status of this flag (tabs or spaces). Something like a warning icon that says "hey, you have chosen to use spaces in the preferences, but this file uses tabs, so Atom will indent your lines with tabs". The current behaviour is far from intuitive.

**Comment by izuzak** [comment]
> I would love to have some status bar icon showing the current status of this flag (tabs or spaces).

Please feel free to open an issue in the status bar package: https://github.com/atom/status-bar

**Comment by mark-hahn** [comment]
> I would love to have some status bar icon showing the current status of
this flag

I think @leedohm has a package for this.  He has two or three packages to
help with the tab/space problems.

**Comment by axebox** [comment]
This isn't working for me. If I enable soft tabs I expect that to be the default. I shouldn't have to set tabs for every file I work with.

**Comment by KarlJakober** [comment]
also broken for me, very annoying.

**Comment by Kern0** [comment]
I agree this bug :(

**Comment by ck3g** [comment]
Same for me

**Comment by SlimeQ** [comment]
same for me, makes writing python a nightmare
the checkbox in preferences is extremely misleading, sort of got it right by changing config
still, leading spaces are automatically grouped by 4, making it impossible to change my indent level to 2. this is weird behaviour to be sure

**Comment by Kern0** [comment]
@izuzak i think, we should reopen this.

**Comment by weichert** [comment]
This is a misfeature!
Soft tabs should never be the default behavior. At the very least tell us in the status line what's going on. Even if it's trying to auto-detect, this is too much magic. People rely on correct whitespace like they rely on correct encoding.
Changing user input like this is bad especially if it doesn't even announce itself.
I just learned about this while a heredoc was breaking and now I wonder how many files I may have contaminated.
Absolutely ridiculous!

**Comment by m-antis** [comment]
Set `softTabs` in `config.cson` to `true`

```ruby
"*":
    softTabs: true
``` 

don't forget to reboot Atom

**Comment by dangerousdan** [comment]
I agree that there is still an issue here. We have coding standard at work that requires us to use soft tabs for new/refactored code. Unfortunately, we work with legacy code which is currently littered with hard tabs which is auto-detected and changes my settings. I don't want to ever write a hard tab.

**Comment by mediabeastnz** [comment]
I added softTabs: true to my config and that seem to do the trick. When i view my preferences though it says SoftTabs are disabled? If the config is true shouldn't this setting be checked?

**Comment by mymainmanbrown** [comment]
I also had to set softTabs in my config.cson file to get softTabs to actually be enabled/working.  Is there some other configuration that might be intercepting my editor settings?  In the editor settings in the UI, I have softTabs enabled and a tab length of 2.  The tab length is respected when tabs are inserted, but the soft tab setting is not respected.  It is only by putting it in my cson file that I can truly have softTabs.

**Comment by izuzak** [comment]
Thanks for all the comments, everyone. https://github.com/atom/atom/pull/8356 introduced a new config setting `editor.tabType` which defines which character is inserted during Tab keypress. Can you all take a look at that and give that a try and see if it resolves your problems? If not, I think we should open a new issue which clearly explains the problem and gives exact steps to reproduce the problem (exact file being opened, exact config settings being used, etc). Does that sound good to everyone?

> I added softTabs: true to my config and that seem to do the trick. When i view my preferences though it says SoftTabs are disabled? If the config is true shouldn't this setting be checked?

@mediabeastnz I wasn't able to reproduce this. Can you open a new issue in https://github.com/atom/settings-view with clear steps to reproduce and [more information](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#submitting-issues)?

**Comment by nanohard** [comment]
I had noticed that even when Soft Tabs is used, it doesn't actually "expand tabs to spaces" as in other editors (awesome, I can use backspace instead of shift+tab again!). This may be where some confusion is happening. I had to enable Show Invisibles to be sure that they were spaces; once I was satisfied I disabled Show Invisibles.

To be clear, this was not the original poster's problem- those were clearly tabs.
Just installed Atom 1.5.3 and Soft Tabs are working well. Time to delete NetBeans, Aptana, Komodo, and several others. I think this issue can be closed.

**Comment by davitb** [comment]
I'm very annoyed with this issue. It's a basic feature for an editor. 
Was looking forward to switch from Sublime to Atom but I can't figure how to make it to insert spaces instead of Tabs.
In my settings I'm using SoftTabs with Tab=4 spaces. But Atom never inserts spaces.

**Comment by hawsome** [comment]
@izuzak here are my steps to reproduce:

1. Check 'Soft Tabs' in 'Settings'.
1. Set 'Tab Type' to `soft` in 'Settings'.
1. Make sure they're set correctly in `config.cson`.
1. Be sad every time I try to tab in all sorts of files (`html`, `js`) it uses hard tabs instead of soft tabs.

I can't understand why this happens, it's like probably the largest bug in the history of Atom. It almost makes me want to switch back to Sublime it's that bad 😢

**Comment by bjenkins101** [comment]
Why is this closed?

**Comment by rahilwazir** [comment]
This issue still exist!

**Comment by dev-xdyang** [comment]
This issue still exist!

**Comment by rasm200** [comment]
This is my first time using Atom. When I tried to use emmet I faced the same problem that every one here, so, after trying with different options, at the end I found that using the keys combination ctrl+shift+enter is more confortable than traditional way, y recomended it, you will be seeing in real time how html will be rendered

**Comment by jeffmcneill** [comment]
Most reports of issues seem to be closed without resolving the underlying problem. There is only one open I could find (on cursory clicking):

## Closed

- #5497 
- #3719
- #5192

## Open 

- #4054

The big issue as reported is the inaccurate "guessing" combined with the inability for an editor-wide setting (and not per-file setting). It may be that several preferences may be required instead of having one monolithic setting which is clearly causing unhappiness:

- Default for all files (y/n)
- Default for certain files (by file extension)
- Setting on per file basis

The idea would be to be able to re-order these in terms of priority, based on the needs of the user.

**Comment by tetreault** [comment]
Please keep this open, its been a while since June and this is still a problem and I agree its about to make me switch away from atom its that annoying...

**Comment by jkw** [comment]
Thought I was alone with this issue, but then I found this issue.
It isn't just the editor itself, seems to be some underlying code that causes it because not even the package tabs-to-spaces can convert tabs in a html-file to pure spaces.

I sure wish there was a never-ever-ever-ever use tabs ever in any file ever regarding of package settings thing to turn **ON** and then delete the code that shows the button.

**Comment by JustinJKlein** [comment]
I have been struggling with this for a long time. I cannot get atom to do a soft tab with 2 spaces. It never never never does it. Incredibly friustrating to go through and set it in the editor and then follow this message chain and not find an answer that actually works. This sucks and is totally a reason to throw this software into my trash.

**Comment by skizot722** [comment]
I was having this exact problem.  Cursor would jump over like it was a hard tab, even find and replace using an actual tab would detect tabs all over the file.  Then I found **Atomic Soft Tabs** option under Editor settings.  It was ticked.  I unticked it and now my problem seems to be solved.  Very frustrating though.

**Comment by tetreault** [comment]
I solved this problem by switching to Vim 3 months ago. haven't had a tab/space issue or my text editor crash yet! 👍💯🔥

**Comment by miikaah** [comment]
Also make sure you've specified the correct indentation style for the file type if you are using .editorconfig .

eg:
`[*]
indent_style = tab

[*.yml]
indent_style = space
indent_size = 2`

**Comment by brienwashburn** [comment]
+1 for skizot722 a couple comments up. Make certain that the option "Atomic Soft Tabs" is unchecked in Editor. I realized something was wrong when I ran the command (using Vim) 

    :%s/\t/1/g 

and nothing was replaced. I ran 

     :%s/ /1/g

and all the supposed tabs turned into 4 1's (I have tabwidth set to 4).  Now when I hit delete on macOS it will actually go a space at a time.

**Comment by spencer1573** [comment]
![mza2xi9](https://user-images.githubusercontent.com/4762850/34076075-f9e9f7a2-e297-11e7-9246-f91eb6e2f749.jpg)

**Comment by ptytb** [comment]
Finally, all pieces together:

Editor / Atomic Soft Tabs = Off
Editor / Soft Tabs = On
Editor / Tab Length = 4
Editor / Tab Type = auto
Install **auto-detect-indentation** plugin
Replace existing tabs with spaces :%s/\t/    /g

Works like a charm.
Plus, _Alt+Backspace Tab_ works like "==" (auto indent) in vim mode.

**Comment by timsully** [comment]
@ptytb When you say "Replace existing tabs with spaces :%s/\t/ /g" where do I go to add this **:%s/\t/ /g** ? Or am I not understanding this correctly.

**Comment by ptytb** [comment]
@timsully this command is supposed to be used in Vim mode (*vim-mode-plus* plugin). It has the same effect as hitting Ctrl-Shift-F and replacing space with \t in regexp mode.
